### ISSUE TITLE
I can't insert a new Computer CI.

### ISSUE THREAD

**Issue by prighi61** [code4me/4me-sdk-csharp#33]
Hi Guys,

I can't insert a new CI, I always get "Product is invalid". I tried even duplicating an existing CI into a new CI in this way:
ConfigurationItem ci1;
ConfigurationItem ci = new ConfigurationItem();
FilterCollection filterCI1 = new FilterCollection();
filterCI1.Add(new Filter("Label", FilterCondition.Equality, "CMP00079"));
List resultci = client.ConfigurationItems.Get(filterCI1, "*");
if (resultci.Count.Equals(1))
{
ci1 = resultci[0];

                    // Crea CI-Computer in 4me
                    ci.AssetID = ci1.AssetID;
                    ci.DepreciationMethod = ci1.DepreciationMethod;
                    ci.FinancialOwner = ci1.FinancialOwner;
                    ci.Label = ci1.Label + "-1";
                    ci.LicenseExpiryDate = ci1.LicenseExpiryDate;
                    ci.LicenseType = ci1.LicenseType;
                    ci.Location = ci1.Location;
                    ci.Name = ci1.Name;
                    ci.NumberOfCores = ci1.NumberOfCores;
                    ci.NumberOfLicenses = ci1.NumberOfLicenses;
                    ci.NumberOfProcessors = ci1.NumberOfProcessors;
                    ci.PoNr = ci1.PoNr;
                    ci.Product = ci1.Product;
                    ci.PurchaseValue = ci1.PurchaseValue;
                    ci.Rate = ci1.Rate;
                    ci.Remarks = ci1.Remarks;
                    ci.SalvageValue = ci1.SalvageValue;
                    ci.SerialNr = ci1.SerialNr;
                    ci.Service = ci1.Service;
                    ci.Site = ci1.Site;
                    ci.SiteLicense = ci1.SiteLicense;
                    ci.Source = ci1.Source;
                    ci.SourceID = ci1.SourceID + "-1";
                    ci.StartDate = ci1.StartDate;
                    ci.Status = ci1.Status;
                    ci.Supplier = ci1.Supplier;
                    ci.SupportTeam = ci1.SupportTeam;
                    ci.SystemID = ci1.SystemID;
                    ci.TemporaryLicense = ci1.TemporaryLicense;
                    ci.UsefulLife = ci1.UsefulLife;
                    ci.WarrantyExpiryDate = ci1.WarrantyExpiryDate;

                    client.ConfigurationItems.Insert(ci);

but with no success.
Has someone succedeed in inserting a new CI? Could he tell me what I'm doing wrong?
Thanks.

Paolo Righi.

**Comment by klaasvandeweerdt** [comment]
@prighi61, you should have more information in the detailed message of the exception.
I just ran follow code and it works for me
```
                Product product = client.Products.Get(1);

                ConfigurationItem ci = new ConfigurationItem()
                {
                    Product = product,
                    Name = "CI via API 001",
                    Label = "CI via API 001",
                    SerialNr = "CIAPISN001",
                    Status = ConfigurationItemStatusType.BeingRepaired
                };
                client.ConfigurationItems.Insert(ci);
```
Ran a piece of code similar to yours and that works as well
```
                ConfigurationItem ci = new ConfigurationItem();
                ConfigurationItem ci1 = client.ConfigurationItems.Get(new Filter("SerialNr", FilterCondition.Equality, "SerialNumber"), "*")[0];
                ci.AssetID = ci1.AssetID;
                ci.DepreciationMethod = ci1.DepreciationMethod;
                ci.FinancialOwner = ci1.FinancialOwner;
                ci.Label = ci1.Label + "-1";
                ci.LicenseExpiryDate = ci1.LicenseExpiryDate;
                ci.LicenseType = ci1.LicenseType;
                ci.Location = ci1.Location;
                ci.Name = ci1.Name;
                ci.NumberOfCores = ci1.NumberOfCores;
                ci.NumberOfLicenses = ci1.NumberOfLicenses;
                ci.NumberOfProcessors = ci1.NumberOfProcessors;
                ci.PoNr = ci1.PoNr;
                ci.Product = ci1.Product;
                ci.PurchaseValue = ci1.PurchaseValue;
                ci.Rate = ci1.Rate;
                ci.Remarks = ci1.Remarks;
                ci.SalvageValue = ci1.SalvageValue;
                ci.SerialNr = ci1.SerialNr + "-1";
                ci.Service = ci1.Service;
                ci.Site = ci1.Site;
                ci.SiteLicense = ci1.SiteLicense;
                ci.Source = ci1.Source;
                ci.SourceID = ci1.SourceID + "-1";
                ci.StartDate = ci1.StartDate;
                ci.Status = ci1.Status;
                ci.Supplier = ci1.Supplier;
                ci.SupportTeam = ci1.SupportTeam;
                ci.SystemID = ci1.SystemID;
                ci.TemporaryLicense = ci1.TemporaryLicense;
                ci.UsefulLife = ci1.UsefulLife;
                ci.WarrantyExpiryDate = ci1.WarrantyExpiryDate;
                client.ConfigurationItems.Insert(ci);
```
So I expect that one of the fields is missing or not unique, for example serial has to be unique.

**Comment by prighi61** [comment]
Hi Klaas,

You are right (I replaced id=1 with id=122 as id=1 doesn’t exist). I can Insert new CI using already existant Products.
I created new products: using the following new Product I can’t insert a new CI:
Name
Value
Type
◢
Existant Product
122
Sdk4me.Product
Brand
"VMware"
string
Category
"computer/virtual_machine"
string
▶ CreatedAt
{5/7/2020 1:32:51 PM}
System.DateTime?
DepreciationMethod
NotApplicableCostIsZero
Sdk4me.ConfigurationItemDepreciationMethodType?
Disabled
false
bool
▶ FinancialOwner
6
Sdk4me.Organization
FinancialOwnerID
6
long?
ID
122
long
▶ IncludeDuringSerialization
Count = 0
System.Collections.Generic.HashSet<string>
Model
"ESXi"
string
Name
"VMware ESXi Virtual Machine"
string
PictureUri
"https://4me-demo-assets.s3-accelerate.amazonaws.com/avatars/products/original/logo-vmware2.svg"
string
ProductID
null
string
Rate
null
float?
Remarks
null
string
RuleSet
Server
Sdk4me.ProductRuleSetType?
▶ Service
42
Sdk4me.Service
ServiceID
42
long?
Source
null
string
SourceID
null
string
▶ Supplier
6
Sdk4me.Organization
SupplierID
6
long?
▶ SupportTeam
16
Sdk4me.Team
SupportTeamID
16
long?
UIExtension
null
Sdk4me.UIExtension
UIExtensionID
null
long?
▶ UpdatedAt
{5/19/2020 9:13:18 AM}
System.DateTime?
UsefulLife
null
int?

Name
Value
Type
◢
New Product
268
Sdk4me.Product
Brand
"Xen"
string
Category
"computer/desktop_pc_or_workstation"
string
▶ CreatedAt
{5/13/2020 1:50:34 PM}
System.DateTime?
DepreciationMethod
null
Sdk4me.ConfigurationItemDepreciationMethodType?
Disabled
false
bool
FinancialOwner
null
Sdk4me.Organization
FinancialOwnerID
null
long?
ID
268
long
▶ IncludeDuringSerialization
Count = 0
System.Collections.Generic.HashSet<string>
Model
"HVM domU"
string
Name
"Xen HVM domU"
string
PictureUri
null
string
ProductID
null
string
Rate
null
float?
Remarks
null
string
RuleSet
PhysicalAsset
Sdk4me.ProductRuleSetType?
▶ Service
31
Sdk4me.Service
ServiceID
31
long?
Source
"BCMComputerSystem"
string
SourceID
"Xen HVM domU"
string
Supplier
null
Sdk4me.Organization
SupplierID
null
long?
▶ SupportTeam
11
Sdk4me.Team
SupportTeamID
11
long?
UIExtension
null
Sdk4me.UIExtension
UIExtensionID
null
long?
▶ UpdatedAt
{5/18/2020 3:50:13 PM}
System.DateTime?
UsefulLife
null
int?

Why the newly insert product is invalid?
Thank you a lot for your assistance.
Best Regards.
Paolo Righi
Product Specialist

[cid:image001.jpg@01D62DBE.EF3F5670]
C.H. Ostfeld
T: +39 0266800303

prighi@ostfeld.com<mailto:prighi@ostfeld.com> - www.ostfeld.com<http://www.ostfeld.com/>

Viale Stelvio 70 - 20159, Milano

[cid:image002.jpg@01D62DBE.EF3F5670]<https://www.linkedin.com/company/c-h-ostfeld>   [cid:image003.jpg@01D62DBE.EF3F5670] <https://twitter.com/chostfeld>    [cid:image004.jpg@01D62DBE.EF3F5670] <https://www.ostfeld.com/#sendgrid_nlvx_widget-4>

**Comment by prighi61** [comment]
Hi Klaas,

I forgot to mention that I could create a new CI via the 4me console using Product id=268:
Name
Value
Type
◢
ci4
2483
Sdk4me.ConfigurationItem
AssetID
null
string
▶ CreatedAt
{5/18/2020 12:21:25 PM}
System.DateTime?
▶ CustomFields
Count = 0
Sdk4me.CustomFieldCollection
▶ CustomFieldsPrivate
Count = 0
System.Collections.Generic.List<Sdk4me.CustomField>
DepreciationMethod
null
Sdk4me.ConfigurationItemDepreciationMethodType?
FinancialOwner
null
Sdk4me.Organization
FinancialOwnerID
null
long?
ID
2483
long
▶ IncludeDuringSerialization
Count = 0
System.Collections.Generic.HashSet<string>
Label
"CMP00079"
string
LicenseExpiryDate
null
System.DateTime?
LicenseType
null
Sdk4me.ConfigurationitemLicenseType?
Location
null
string
Name
"Xen HVM domU"
string
NumberOfCores
null
int?
NumberOfLicenses
null
int?
NumberOfProcessors
null
int?
PoNr
null
string
▶ Product
268
Sdk4me.Product
ProductID
268
long?
PurchaseValue
null
float?
Rate
null
float?
Remarks
null
string
RuleSet
PhysicalAsset
Sdk4me.ConfigurationItemRuleSetType?
SalvageValue
null
float?
SerialNr
"ASDASD"
string
▶ Service
31
Sdk4me.Service
ServiceID
31
long?
Site
null
Sdk4me.Site
SiteID
null
long?
SiteLicense
null
bool?
Source
"4me"
string
SourceID
null
string
StartDate
null
System.DateTime?
Status
InProduction
Sdk4me.ConfigurationItemStatusType?
Supplier
null
Sdk4me.Organization
SupplierID
null
long?
▶ SupportTeam
11
Sdk4me.Team
SupportTeamID
11
long?
SystemID
null
string
TemporaryLicense
null
bool?
▶ UpdatedAt
{5/18/2020 12:21:25 PM}
System.DateTime?
UsefulLife
null
int?
WarrantyExpiryDate
null
System.DateTime?

So, why can’t I create a new CI using product id=268?

Best Regards.
Paolo Righi
Product Specialist

[cid:image001.jpg@01D62DC1.B4444780]
C.H. Ostfeld
T: +39 0266800303

prighi@ostfeld.com<mailto:prighi@ostfeld.com> - www.ostfeld.com<http://www.ostfeld.com/>

Viale Stelvio 70 - 20159, Milano

[cid:image002.jpg@01D62DC1.B4444780]<https://www.linkedin.com/company/c-h-ostfeld>   [cid:image003.jpg@01D62DC1.B4444780] <https://twitter.com/chostfeld>    [cid:image004.jpg@01D62DC1.B4444780] <https://www.ostfeld.com/#sendgrid_nlvx_widget-4>

**Comment by prighi61** [comment]
Hi @klaasvandeweerdt ,

I tried this:
                    ConfigurationItem ci = new ConfigurationItem();
                    FilterCollection filterCI1 = new FilterCollection();
                    filterCI1.Add(new Filter("Label", FilterCondition.Equality, "CMP00079"));
                    List<ConfigurationItem> resultci = client.ConfigurationItems.Get(filterCI1, "*");
                    if (resultci.Count.Equals(1))
                    {
                        ConfigurationItem ci1 = resultci[0];

                        ci.AssetID = ci1.AssetID;
                        ci.DepreciationMethod = ci1.DepreciationMethod;
                        ci.FinancialOwner = ci1.FinancialOwner;
                        ci.Label = ci1.Label + "-1";
                        ci.LicenseExpiryDate = ci1.LicenseExpiryDate;
                        ci.LicenseType = ci1.LicenseType;
                        ci.Location = ci1.Location;
                        ci.Name = ci1.Name;
                        ci.NumberOfCores = ci1.NumberOfCores;
                        ci.NumberOfLicenses = ci1.NumberOfLicenses;
                        ci.NumberOfProcessors = ci1.NumberOfProcessors;
                        ci.PoNr = ci1.PoNr;
                        ci.Product = ci1.Product;
                        ci.PurchaseValue = ci1.PurchaseValue;
                        ci.Rate = ci1.Rate;
                        ci.Remarks = ci1.Remarks;
                        ci.SalvageValue = ci1.SalvageValue;
                        ci.SerialNr = ci1.SerialNr + "-1";
                        ci.Service = ci1.Service;
                        ci.Site = ci1.Site;
                        ci.SiteLicense = ci1.SiteLicense;
                        ci.Source = ci1.Source;
                        ci.SourceID = ci1.SourceID + "-1";
                        ci.StartDate = ci1.StartDate;
                        ci.Status = ci1.Status;
                        ci.Supplier = ci1.Supplier;
                        ci.SupportTeam = ci1.SupportTeam;
                        ci.SystemID = ci1.SystemID;
                        ci.TemporaryLicense = ci1.TemporaryLicense;
                        ci.UsefulLife = ci1.UsefulLife;
                        ci.WarrantyExpiryDate = ci1.WarrantyExpiryDate;

                        client.ConfigurationItems.Insert(ci);

but I get the same error: "Product is invalid".
"CMP00079" is a new CI created by the 4me console that uses an "invalid" product.

Best Regards.

**Comment by mathijssterk** [comment]
@prighi61 Not sure if this is the case, but one thing to keep in mind when creating new CIs using the 4me API is the following:

To facilitate integrations with discovery tools, the POST is treated as a PATCH in case the provided name or label is already used by an inactive CI in the same account.

When a discovery tool can no longer find a certain CI, it will report that and in 4me that CI will be set to the status removed. When that CI reappears some time later, discovery tools will treat this CI as new, leading to a POST request to 4me with the same name/label. In that case an error will be provided by 4me stating name/label is already taken. To prevent this error, special handling was provided to interpret the POST as a PATCH of the existing inactive CI.

Manually testing discovery tool integrations, it sometimes happens that you use a name/label that you’ve previously used for a CI that was already inactive. Normally this is not a problem, as 4me will update that existing inactive CI and you will see the results that you are looking for.

But in some cases the new data sent with the CI POST request will conflict with the existing data known about the (removed) CI in 4me. Then you may get unexpected errors.

A simple solution you can use during development of the integration is to ensure that the name and label fields are always unique by adding a random postfix to it.

**Comment by prighi61** [comment]
Hi @mathijssterk ,
thank you for your hint. I think this is not the case as I'm using a demo and I've never make inactive anything.
However I will mind your suggestion.
Best regards.

**Comment by klaasvandeweerdt** [comment]
@prighi61, here is a small piece of code that works for me.
```                
ProductCategory productCategory = client.ProductCategories.Get(1);
Team team = client.Teams.Get(1);

Product product = new Product
{
    Name = "Custom product",
    RuleSet = ProductRuleSetType.PhysicalAsset,
    Brand = "Custom brand",
    Model = "My custom product model",
    Category = productCategory.Reference
};
product = client.Products.Insert(product);

ConfigurationItem ci = new ConfigurationItem()
{
    Product = product,
    Name = "CI via API 1001",
    Label = "CI via API 1001",
    SerialNr = "CIAPISN1001",
    Status = ConfigurationItemStatusType.BeingRepaired,
    SupportTeam = team
};
ci = client.ConfigurationItems.Insert(ci);
```
When using an insert or update method make sure to use the return value of the method, the object passed as the argument will remain unchanged.

In this case the product id remains 0.
```
client.Products.Insert(product);
```
The only way to get the newly created product is like this
```
product = client.Products.Insert(product);
```
Hopes this helps

**Comment by prighi61** [comment]
Hi Guys,

I want to let you know that I'm facing the same problems via curl commands. So this library in not implied.
Thank you anyway for your suggestions.
Paolo Righi.

**Comment by prighi61** [comment]
Hi Guys,

I want to let you know that 4me support can reproduce this issue.
A 4me support team member's answer:
"I identified that this problem only occurs within a 4me Demo instance and then only when posting new CI records using the 4me API when the CI posted also references a new Product record.
I have escalated this issue to our development team for resolution".
I'm now waiting for the resolution.

Paolo Righi.
